core/tx_pool.go

```diff
@@ -680,6 +685,29 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (bool, error) {
 			pool.removeTx(tx.Hash(), false)
 		}
 	}
+	// If there isn't enough free size in pool, discard underpriced transactions
+	if pool.all.TotalSize() + tx.Size() > pool.config.TotalPoolSizeLimit {
```

Comment:
You can't lock in add, you already have a lock. Please remove the TotalSize method altogether and just directly access the field.